presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HistoricalStatisticsEquivalentPlanMarkingOptimizer.java

```diff
@@ -55,13 +57,36 @@ public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, Pla
     private static class Rewriter
             extends SimplePlanRewriter<Void>
     {
+        private final Stack<Long> limits = new Stack<>();
+        private final PlanNodeIdAllocator idAllocator;
+
+        public Rewriter(PlanNodeIdAllocator idAllocator)
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+        }
+
         @Override
         public PlanNode visitPlan(PlanNode node, RewriteContext<Void> context)
         {
             if (!node.getStatsEquivalentPlanNode().isPresent()) {
-                node = node.assignStatsEquivalentPlanNode(Optional.of(node));
+                PlanNode nodeWithLimit = node;
+                // A LIMIT node can affect statistics of several children nodes, so let's put the limit node in the stats equivalent plan node.
+                // This way we always include the limit when hashing each subplan.
```

Comment:
Not really getting the idea behind this .... Could you elaborate more? In general, we do't put a global variable like `limits` in this case. We embed it into the `context`. But even that, would be good to have example or clear comments.

Reply:
Updated comment